bundles/specmate-dbprovider-h2/src/specmate/dbprovider/h2/H2Provider.java

```diff
@@ -0,0 +1,172 @@
+		"service.ranking:Integer=2" })
+public class H2Provider implements IDBProvider {
+
+	private ConfigurationAdmin configurationAdmin;
+	private Connection connection;
+	private String jdbcConnection;
+	private String repository;
+	private String resource;
+	private String userResource;
+	private boolean isVirginDB;
+	private Pattern databaseNotFoundPattern = Pattern.compile(".*Database \\\".*\\\" not found.*", Pattern.DOTALL);
+
+	@Activate
+	public void activate() throws SpecmateException {
+		this.isVirginDB = false;
+		readConfig();
+	}
+
+	@Override
+	public void readConfig() throws SpecmateException {
+		try {
+			Dictionary<String, Object> configProperties = configurationAdmin.getConfiguration(H2ProviderConfig.PID)
+					.getProperties();
+			this.jdbcConnection = (String) configProperties.get(H2ProviderConfig.KEY_JDBC_CONNECTION);
+			this.repository = (String) configProperties.get(H2ProviderConfig.KEY_REPOSITORY_NAME);
+			this.resource = (String) configProperties.get(H2ProviderConfig.KEY_RESOURCE_NAME);
+			this.userResource = (String) configProperties.get(H2ProviderConfig.KEY_USER_RESOURCE_NAME);
+
+			String failmsg = " not defined in configuration.";
+			if (StringUtils.isNullOrEmpty(this.jdbcConnection)) {
+				throw new SpecmateException("JDBC connection" + failmsg);
+			}
+
+			if (StringUtils.isNullOrEmpty(this.repository)) {
+				throw new SpecmateException("Database repository" + failmsg);
+			}
+
+			if (StringUtils.isNullOrEmpty(this.resource)) {
+				throw new SpecmateException("Database resource" + failmsg);
+			}
+
+			if (StringUtils.isNullOrEmpty(this.userResource)) {
+				throw new SpecmateException("Database user resource" + failmsg);
+			}
+		} catch (IOException e) {
+			throw new SpecmateException("Could not obtain database configuration.", e);
+		}
+	}
+
+	@Override
+	public Connection getConnection() throws SpecmateException {
+		if (this.connection == null) {
+			try {
+				initiateDBConnection();
+				// In development, when specmate or the tests are run for the
+				// first time, no database exists (neither on the
+				// file system, nor in memory). There is no sane way to check if
+				// a database exists, except by connecting
+				// to it. In case it does not exist, an SQL exception is thrown.
+				// While in all other possible error cases
+				// we want the client to handle the error, in the situation that
+				// the database does not exist, we want
+				// specmate to continue, without performing a migration, because
+				// the next step CDO performs is to create the database.
+			} catch (SpecmateException e) {
+				Matcher matcher = databaseNotFoundPattern.matcher(e.getCause().getMessage());
+				if (matcher.matches()) {
+					this.isVirginDB = true;
+				} else {
+					throw e;
+				}
+			}
+		}
+
+		return connection;
+	}
+
+	@Override
+	public boolean isVirginDB() throws SpecmateException {
+		if (this.connection == null) {
```

Comment:
Can this lead to leaking open connections? You wont expect a call to isVirginDb to open a connection. Maybe better check if connection is present and otherwise throw an exception

Reply:
I don't think it would leak a connection. The method getConnection makes sure of that by only creating connections when the member variable is null.
However, to avoid future bugs, your suggestion makes sense. Fixed.